text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,278 @@
+Other dependencies of `std` besides core we don't need to worry about, because they are only transitive dependencies through `std`, not direct dependencies.
+
+Now, not all crates depend on these crates, so there must be a way to opt out.
+The primary way is just to create a conflict.
+If an (explicit) dependency has the same name as one of the implicit defaults, implicit dependencies of the same sort will be skipped.
+For example, if a crate explicit depends on `std` as a regular dependency, neither `std` nor any other implicit regular dependency will be injected.
+Since currently regular dependencies can be included stdlib dependencies, only regular dependencies can be opted out of.
+This means we are free to change the implicit dev and build dependencies without breaking anything.
+
+Opting out via conflict as described above is adequate for almost all cases.
+The one exception is `core` itself, which must of course not depend on `core` or `std` implicitly or explicitly---or anything else for that matter.
+For it, a key, `implicit-dependencies = <true|false>`, will be introduced.
+Because it doesn't generalize if we make the implicit build or dev stdlib deps optional, this key will be permanently unstable.
+
+## Compiler language version
+
+Compilers besides rustc may have version numbers distinct from the version of Rust they implement.
+For this purpose, the verbose version output (`$CARGO_RUSTC -vV`) should contain an additional line:
+```
+language-version: <version>
+```
+For now, `<version>` should be a version, not a version requirement, and the patch number must be zero as patch numbers don't make sense for interfaces.
+[This may be relaxed in the future for compilers which implement multiple versions.]
+
+## Compiler Source
+
+While the standard library *interface* is defined with each rustc version, the implementation of many crates, by virtue of using unstable features, is compiler-specific.
+This makes the standard library unfit for crates.io.
+(Additionally, the issue of dealing with nightly also makes crates.io hard to use, but that is a less clear-cut obstacle.)
+
+To get around this, Cargo will give compilers the option of distributing the source of their implementation of the standard library in a location Cargo knows of.
+Conveniently, Cargo has a "source" abstraction for providers of packages.
+Examples of this are file-system paths, git repositories, and the upcoming registries.
+To implement this, Cargo will gain knowledge of of a new source, the "compiler source".
+The compiler source, if it is present, will be located in the sysroot in `<sysroot>/lib/rustlib/src`.
+The exact format this takes will be determined during implementation and added back to this RFC before stabilization, but that of a "local registry" is likely, now that
+https://github.com/rust-lang/cargo/pull/2857 has landed.
+Compilers should include the source of each crate of their implementation of the standard library in side.
```

Comment:
inside* and best would be if you remove it (the sentence would make more sense that way).